liblarch/filteredtree.py

```diff
@@ -620,10 +620,13 @@ def apply_filter(self, filter_name, parameters=None,
                 raise ValueError(
                     "No filter of name {} in the bank".format(filter_name))
 
-        if filter_name not in self.applied_filters:
-            self.applied_filters.append(filter_name)
-            if refresh:
-                self.refilter()
+        filter_predicate = filter_name not in self.applied_filters or\
```

Comment:
Simplify this code. Put at the begging of the function:

```python
should_refilter = False
```

When setting parameters, line 618
```
if filt:
   filt.set_parameters(parameters)
   should_refilter = True
```

When adding filter_name into applied_filters, do again `should_refilter=True`

Before return of this function, just do
```python
if should_refilter:
     self.refilter()
```

It would result in much less complicated code